Validator does not consider custom heading IDs

### Describe the bug

I use the `remark-custom-heading-id` according to [this guide](https://github.com/withastro/starlight/discussions/2050#discussioncomment-9868037).

Everything works well except the link validation. The validator does not recognize the custom header IDs and fails on links that point to this custom ID.

I also use `rehypeHeadingIds` and `rehypeExternalLinks` plugins.

Here is my `markdown` configuration:

```js
markdown: {
  remarkPlugins: [
    remarkHeadingId, // Support custom heading IDs.
  ],
  rehypePlugins: [
    rehypeHeadingIds,
    [
      rehypeAutolinkHeadings,
      {
        behavior: 'wrap', // Wrap the heading text in a link.
      },
    ],
    [
      rehypeExternalLinks,
      {
        target: '_blank', // Open external links in a new tab.
      }
    ]
  ],
}
```

### To Reproduce

1. Create a heading with the custom ID:

```md
## Heading {#custom-heading}
```

2. Link to this heading:

```md
Check the [custom heading](#custom-heading).
```

3. Try to build the project and see the error:

```console
validating links
11:01:59 [ERROR] ✗ Found 1 invalid link in 1 file.
11:01:59 ▶ docs/
11:01:59   └─ #custom-heading - invalid anchor

[AstroUserError] Links validation failed.
```

### Expected behavior

No validation error.

### How often does this bug happen?

Every time

### System Info

Operating system: MacOS

### Additional Context

_No response_

--- a/.github/workflows/integration.yml
+++ b/.github/workflows/integration.yml
@@ -22,7 +22,7 @@ jobs:
         uses: actions/checkout@v3
 
       - name: Install pnpm
-        uses: pnpm/action-setup@v2.2.4
+        uses: pnpm/action-setup@v4
         with:
           version: 8.6.3
 

--- a/.github/workflows/integration.yml
+++ b/.github/workflows/integration.yml
@@ -19,7 +19,7 @@ jobs:
     runs-on: ${{ matrix.os }}
     steps:
       - name: Checkout
-        uses: actions/checkout@v3
+        uses: actions/checkout@v4
 
       - name: Install pnpm
         uses: pnpm/action-setup@v4
@@ -27,7 +27,7 @@ jobs:
           version: 8.6.3
 
       - name: Install Node.js
-        uses: actions/setup-node@v3
+        uses: actions/setup-node@v4
         with:
           cache: pnpm
           node-version: 18

--- a/.github/workflows/release.yml
+++ b/.github/workflows/release.yml
@@ -25,7 +25,7 @@ jobs:
           fetch-depth: 0
 
       - name: Install pnpm
-        uses: pnpm/action-setup@v2.2.4
+        uses: pnpm/action-setup@v4
         with:
           version: 8.6.3
 

--- a/.github/workflows/release.yml
+++ b/.github/workflows/release.yml
@@ -20,7 +20,7 @@ jobs:
       id-token: write
     steps:
       - name: Checkout
-        uses: actions/checkout@v3
+        uses: actions/checkout@v4
         with:
           fetch-depth: 0
 
@@ -30,7 +30,7 @@ jobs:
           version: 8.6.3
 
       - name: Install Node.js
-        uses: actions/setup-node@v3
+        uses: actions/setup-node@v4
         with:
           cache: pnpm
           node-version: 18

--- a/packages/starlight-links-validator/libs/remark.ts
+++ b/packages/starlight-links-validator/libs/remark.ts
@@ -39,6 +39,11 @@ export const remarkStarlightLinksValidator: Plugin<[base: string], Root> = funct
       // https://github.com/syntax-tree/mdast-util-mdx-jsx#nodes
       switch (node.type) {
         case 'heading': {
+          if (node.data?.hProperties?.['id']) {
+            fileHeadings.push(String(node.data.hProperties['id']))
+            break
+          }
+
           const content = toString(node)
 
           if (content.length === 0) {

--- a/packages/starlight-links-validator/package.json
+++ b/packages/starlight-links-validator/package.json
@@ -31,6 +31,7 @@
     "@types/picomatch": "2.3.3",
     "astro": "4.0.4",
     "mdast-util-mdx-jsx": "3.0.0",
+    "remark-custom-heading-id": "2.0.0",
     "typescript": "5.1.3",
     "unified": "11.0.4",
     "vfile": "6.0.1",

--- a/packages/starlight-links-validator/tests/custom-ids.test.ts
+++ b/packages/starlight-links-validator/tests/custom-ids.test.ts
@@ -1,0 +1,17 @@
+import { expect, test } from 'vitest'
+
+import { ValidationErrorType } from '../libs/validation'
+
+import { expectValidationErrorCount, expectValidationErrors, loadFixture } from './utils'
+
+test('should validate links with custom IDs', async () => {
+  expect.assertions(2)
+
+  try {
+    await loadFixture('custom-ids')
+  } catch (error) {
+    expectValidationErrorCount(error, 1, 1)
+
+    expectValidationErrors(error, 'test/', [['#heading-with-custom-id', ValidationErrorType.InvalidAnchor]])
+  }
+})

--- a/packages/starlight-links-validator/tests/fixtures/custom-ids/astro.config.ts
+++ b/packages/starlight-links-validator/tests/fixtures/custom-ids/astro.config.ts
@@ -1,0 +1,17 @@
+import starlight from '@astrojs/starlight'
+import { defineConfig } from 'astro/config'
+import remarkCustomHeadingId from 'remark-custom-heading-id'
+
+import starlightLinksValidator from '../..'
+
+export default defineConfig({
+  markdown: {
+    remarkPlugins: [remarkCustomHeadingId],
+  },
+  integrations: [
+    starlight({
+      plugins: [starlightLinksValidator()],
+      title: 'Starlight Links Validator Tests - custom IDs',
+    }),
+  ],
+})

--- a/packages/starlight-links-validator/tests/fixtures/custom-ids/src/content/docs/test.md
+++ b/packages/starlight-links-validator/tests/fixtures/custom-ids/src/content/docs/test.md
@@ -1,0 +1,12 @@
+---
+title: Test
+---
+
+# Heading
+
+# Heading with custom ID {#custom-id}
+
+- [Link to heading with no custom ID](#heading)
+
+- [Link to heading with custom ID](#custom-id)
+- [Link to heading with custom ID using original ID](#heading-with-custom-id)

